src/EnergyPlus/ElectricPowerServiceManager.cc

```diff
@@ -3512,6 +3512,12 @@ ElectricStorage::ElectricStorage( // main constructor
                 }
             }
 
+            if (UtilityRoutines::SameString(state.dataIPShortCut->cAlphaArgs(8), "CurrentAndEnergy")) {
+                conserveBattery_ = BatteryEnergyConservationType::CurrentAndEnergy;
+            } else {
+                conserveBattery_ = BatteryEnergyConservationType::Basic;
+            }
```

Comment:
If you have an enum you should just look it up with getEnumerationValue (soon to be getEnumValue) using a new constexpr string_view array. But that may not be necessary since I don't like these options. Basic does not conserve energy nor does CurrentAndEnergy. So you will have to explain why conservation of energy is not possible with this model. I have not looked at the charge and discharge curves but I would expect a SOC and voltage relationship. Why can't a charge or discharge cycle know exactly how much energy was transferred? @EnergyArchmage may know more about this model?

Reply:
I will try my best to explain what I am seeing in this model as a potential issue.  During charging, the model calculates a new level of "state of charge" in amp-hours based on the current.  It's got two pots--one for available and one for bound.  The sum of those two equals the amount stored in the battery.  Based on the current and voltage, it also calculates how much energy in Joule is charged or discharged using basically a form of current times voltage.  The voltage is a function of the fraction charged.  During discharging, the same is basically true except that there are two correlations for voltage--one for charging and one for discharging.  If the voltage calculated during discharging is higher than the voltage during charging at the same state of charge and current, then it is possible that since energy is based on current times voltage that the discharging energy is greater than the charging energy to get to that point, purely because the correlation between fraction charged and voltage result in different values.  I am tracking both amp-hours and current in this attempt at a fix and zeroing out amp-hours when energy gets to zero.  This avoids taking out more than one puts in.  There are potentially other approaches--like looking at the charging and discharging curves, comparing the two and notifying the user if they might run into problems and what the problem would look like.  I realize that I failed to add additional documentation to explain this when I pushed this and turned it into a pull request.  I can definitely correct that if we stick with what I proposed.  However, if there are still concerns about that and we want to go with an approach that looks at the curve up front or something else, I am open to discussing that.

Reply:
It does sound like bad input curves are the problem here.  Checking the curves up front seems like a helpful thing to do.  

I wonder if the output variable Electric Storage Total Voltage could be examined during times discharging and compared to times when charging to identify/explain the problem.

Reply:
Ok, it seems like it a bad input curve problem.  I looked at the voltage when in charging mode.  The max value is 33.8 V and the average (when not zero) is 33.3 V.  For discharging, the max value is 39 V and the average is 38.7 V.  Since discharging voltage is always higher than charging, that's where it is creating energy.  So, perhaps a better approach would be to check the curves after and notify the users when discharging is higher than charging.  I can create a subroutine that goes through a range of values and compares the two.  If discharging is higher than charging, then a warning would be produced.  I can leave the new output variables in there, but get rid of the new input and the code that processes it and then uses it.  Before I move forward with that, I'd like to hear back that yes that seems reasonable from both @rraustad and @EnergyArchmage.  That way, I'm getting it right (the second time).  Thanks!

Reply:
That seems reasonable to me.

Reply:
Yes that seems reasonable. And you can get rid of the new input field for Battery Conservation Method. I'm not sure if it's useful to have a starting SOC input as Initial Energy Stored in Battery but keeping or deleting that input field is up to you. What was the original assumption for SOC?

Reply:
There is an input for SOC in amp-hours.  I think I will just get rid of the extra bookkeeping for storage in Joules.  Essentially, the program will now alert the user when discharging voltage is higher than charging voltage and what to do to resolve that.  I'll add some text to the IO Ref to alert the users about that also.  I'll let you both know when I've implemented the new approach and it's ready for another look.  Thanks to both of you for your help on this!

Reply:
I should have also said that I'll get rid of both of the new input fields.

Reply:
Is this a check of the curves in getInput or a recurring check during the simulation? I'd rather see a check of the curves in getInput.

Reply:
Definitely during the getInput routine.  It needs to check different charge fractions so that discharging voltage stays below charging voltage for any charged fraction.  If I do that upfront, then there should not be any need to do it during the actual simulation where it could just end up being expensive overhead over an annual simulation.

Reply:
Ok, this is ready to be looked at again.  Tests are still churning, but I don't anticipate any red flags.  I got rid of what was done before and replaced it with a subroutine that gets called during the get routine.  Goes through a range of values for the charging and discharging curves.  If discharging voltage is higher than charging, it triggers a warning message.  The test file that is already in the test suite, ShopwithPVBattery.idf, actually triggers the warning now, which it should given the values of the parameters.  It's a short run so it doesn't end up showing the discharging energy is higher than charging.  However, if you switch to a full annual run, discharging is higher.  I haven't changed the idf yet.  If we feel that it would be better to have a file that doesn't trip this warning message, I'm ok with that.  Happy reviewing!